### ISSUE TITLE
yolov5 best.pt in dnn

### ISSUE THREAD

**Issue by sidh1603** [ultralytics/yolov5#10664]
Hi , is there any way to use yolov5 weights  with darknet like this
 net = cv2.dnn.readNetFromDarknet(configPath, weightsPath) 

because yolov5 gives us .pt weights , where as dnn need config file and weight file .

Any help??

**Comment by sidh1603** [comment]
This is my code


# import the necessary packages
from skimage.metrics import structural_similarity as compare_ssim
import argparse
import imutils
import cv2

import numpy as np
import time
import os
from PIL import Image
import torch

# construct the argument parse and parse the arguments
ap = argparse.ArgumentParser()
ap.add_argument("-i1", "--image1", required=True,
	help="path to input image")
ap.add_argument("-i2", "--image2", required=True,
	help="path to input image")
ap.add_argument("-y", "--yolo", required=True,
	help="base path to YOLO directory")
ap.add_argument("-c", "--confidence", type=float, default=0.5,
	help="minimum probability to filter weak detections")
ap.add_argument("-t", "--threshold", type=float, default=0.3,
	help="threshold when applyong non-maxima suppression")
args = vars(ap.parse_args())

# load the two input images
imageA = cv2.imread(args['image1'])
imageA = cv2.resize(imageA, (640,640))
imageB = cv2.imread(args['image2'])
imageB = cv2.resize(imageB, (640,640))

grayA = cv2.cvtColor(imageA, cv2.COLOR_BGR2GRAY)
grayB = cv2.cvtColor(imageB, cv2.COLOR_BGR2GRAY)



# compute the Structural Similarity Index (SSIM) between the two
# images, ensuring that the difference image is returned
(score, diff) = compare_ssim(grayA, grayB, full=True)
diff = (diff * 255).astype("uint8")
print("SSIM: {}".format(score))


# threshold the difference image, followed by finding contours to
# obtain the regions of the two input images that differ
thresh = cv2.threshold(diff, 0, 255,
	cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]
cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL,
	cv2.CHAIN_APPROX_SIMPLE)
cnts = imutils.grab_contours(cnts)

crops = []
# loop over the contours
for c in cnts:
    # compute the bounding box of the contour and then draw the
    # bounding box on both input images to represent where the two
    # images differ
    (x, y, w, h) = cv2.boundingRect(c)
    cv2.rectangle(imageA, (x, y), (x + w, y + h), (0, 0, 255), 2)
    cv2.rectangle(imageB, (x, y), (x + w, y + h), (0, 0, 255), 2)
    crops.append(imageB[y : y + h, x : x + w])

# load the COCO class labels our YOLO model was trained on
labelsPath = os.path.sep.join([args["yolo"], "coco.names"])
LABELS = open(labelsPath).read().strip().split("\n")

# initialize a list of colors to represent each possible class label
np.random.seed(42)
COLORS = np.random.randint(0, 255, size=(len(LABELS), 3),dtype="uint8")

# derive the paths to the YOLO weights and model configuration
weightsPath = os.path.sep.join([args["yolo"], "yolov3.weights"])
configPath = os.path.sep.join([args["yolo"], "yolov3.cfg"])

#weightsPath = os.path.sep.join([args["yolo"], "yolo-coco/best (1).pt"])
#configPath = os.path.sep.join([args["yolo"], "yolov7x-custom.cfg"])

# load our YOLO object detector trained on COCO dataset (80 classes)
print("[INFO] loading YOLO from disk...")
net = cv2.dnn.readNetFromDarknet(configPath, weightsPath)
#net = torch.hub.load(weightsPath)


objectname = []
for crop in crops:
    # load our input image and grab its spatial dimensions
    image = crop
    (H, W) = image.shape[:2]

    # determine only the *output* layer names that we need from YOLO
    ln = net.getLayerNames()
    #ln = [ln[i[0] - 1] for i in net.getUnconnectedOutLayers()]
    ln = [ln[i - 1] for i in net.getUnconnectedOutLayers()]

    # construct a blob from the input image and then perform a forward
    # pass of the YOLO object detector, giving us our bounding boxes and
    # associated probabilities
    blob = cv2.dnn.blobFromImage(image, 1 / 255.0, (416, 416),swapRB=True, crop=False)
    #blob = cv2.dnn.blobFromImage(image, 0.00392, (640, 640), (0, 0, 0), True, crop=False)
    #blob = cv2.resize(image, (640, 640), fx=0.5, fy=0.5)
    net.setInput(blob)
    start = time.time()
    layerOutputs = net.forward(ln)
    end = time.time()

    # show timing information on YOLO
    print("[INFO] YOLO took {:.6f} seconds".format(end - start))

    # initialize our lists of detected bounding boxes, confidences, and
    # class IDs, respectively
    boxes = []
    confidences = []
    classIDs = []

    # loop over each of the layer outputs
    for output in layerOutputs:
        # loop over each of the detections
        for detection in output:
            # extract the class ID and confidence (i.e., probability) of
            # the current object detection
            scores = detection[5:]
            classID = np.argmax(scores)
            confidence = scores[classID]

            # filter out weak predictions by ensuring the detected
            # probability is greater than the minimum probability
            if confidence > args["confidence"]:
                # scale the bounding box coordinates back relative to the
                # size of the image, keeping in mind that YOLO actually
                # returns the center (x, y)-coordinates of the bounding
                # box followed by the boxes' width and height
                box = detection[0:4] * np.array([W, H, W, H])
                (centerX, centerY, width, height) = box.astype("int")

                # use the center (x, y)-coordinates to derive the top and
                # and left corner of the bounding box
                x = int(centerX - (width / 2))
                y = int(centerY - (height / 2))

                # update our list of bounding box coordinates, confidences,
                # and class IDs
                boxes.append([x, y, int(width), int(height)])
                confidences.append(float(confidence))
                classIDs.append(classID)

    # apply non-maxima suppression to suppress weak, overlapping bounding
    # boxes
    idxs = cv2.dnn.NMSBoxes(boxes, confidences, args["confidence"],
        args["threshold"])



    # ensure at least one detection exists
    if len(idxs) > 0:
        # loop over the indexes we are keeping
        for i in idxs.flatten():
            # extract the bounding box coordinates
            (x, y) = (boxes[i][0], boxes[i][1])
            (w, h) = (boxes[i][2], boxes[i][3])

            # draw a bounding box rectangle and label on the image
            color = [int(c) for c in COLORS[classIDs[i]]]
            cv2.rectangle(image, (x, y), (x + w, y + h), color, 2)
            text = "{}: {:.4f}".format(LABELS[classIDs[i]], confidences[i])
            objectname.append((LABELS[classIDs[i]], confidences[i]))
            cv2.putText(image, text, (x, y - 10), cv2.FONT_HERSHEY_SIMPLEX,
                0.5, color, 2)
            #print("missing objects is :", objectname)


print("missing objects is :", objectname)
#print(objectname)
# show the output images
cv2.imshow("image1.jpg", imageA)
cv2.imshow("image2.jpg", imageB)
cv2.imshow("Diff.jpg", crop)
#cv2.imshow("Thresh.jpg", thresh)
cv2.waitKey(0)

**Comment by UNeedCryDear** [comment]
opencv could read onnx file
```python
#export
python export.py --weights path/to/weights.pt --img [640,640]  --include onnx
#use
net=cv2.dnn.readNetFromONNX(onnx_path)
size=(640,640,3)
img=np.zeros(size,dtype=np.uint8)
blob=cv2.dnn.blobFromImage(img,1/255.0,size=size[0:2],mean=(0,0,0),swapRB=True)
net.setInput(blob)
out=net.getUnconnectedOutLayersNames()
pred=net.forward(out)
```

**Comment by sidh1603** [comment]
Hi ,but what about yolov5 config file.Sent from my iPhoneOn Jan 4, 2023, at 3:42 AM, UNeedCryDear ***@***.***> wrote:﻿
opencv could read onnx file
python export.py --weights path/to/weights.pt --img [640,640]  --include onnx
 cv2.dnn.readNetFromONNX(onnx_path)

**Comment by UNeedCryDear** [comment]
> Hi ,but what about yolov5 config file

The onnx file contains weights and config files, which can be read directly by ```cv2.dnn.readNetFromONNX('path/to/xxx.onnx')```

**Comment by sidh1603** [comment]
Hi,

Can i convert my yolov5 model code to use CPU instead of GPU.

Kind regards
Siddharth Rana
